### PR TITLE
Fix #6, Fix #7

### LINKED ISSUE #6: Stop command

### LINKED ISSUE #7: Skip command

### PR DESCRIPTION
**Pull request by stuckinab0x**

### CODE REVIEW

**Comment by UncleDave** [comment]
The commit message should still summarise what was done so we don't have to go digging through old issues to figure it out after this is merged. Multiline is fine, lines beyond the first are for the "extended description" if you need one, and I commonly use it for the github stuff.

> Add skip, stop commands
> Fix #6
> Fix #7

You don't need to redo the commit here, we can fix it when the PR is merged.

File: src/bot/commands/skip.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    } else if (skipOption) { // If not Idle and option
+      if (skipOption === 'all') {
+        context.botAudioPlayer.player.stop();
+        context.soundQueue.clear();
+        reply = 'Skipped all sounds. Queue is now empty.';
+      } else if (Number.isInteger(Number(skipOption))) {
+        if (Number(skipOption) > context.soundQueue.length) reply = 'All sounds skipped (count option was >= number of current sounds.)';
+        else reply = `Skipped ${ skipOption } sound(s).`;
+        context.botAudioPlayer.player.stop();
+        const int = Number(skipOption);
```
Comment (UncleDave) [inline src/bot/commands/skip.ts:32]: Can we try to turn `skipOption` into a `Number` only once and reuse it?

File: src/bot/commands/skip.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+      // eslint-disable-next-line no-useless-escape
+      reply = 'No sounds currently playing or in queue! Why not try "/sound limmy are you deaf" ? \:ear_with_hearing_aid: \:smile:';
+    } else if (skipOption) { // If not Idle and option
+      if (skipOption === 'all') {
+        context.botAudioPlayer.player.stop();
+        context.soundQueue.clear();
+        reply = 'Skipped all sounds. Queue is now empty.';
+      } else if (Number.isInteger(Number(skipOption))) {
+        if (Number(skipOption) > context.soundQueue.length) reply = 'All sounds skipped (count option was >= number of current sounds.)';
+        else reply = `Skipped ${ skipOption } sound(s).`;
```
Comment (UncleDave) [inline src/bot/commands/skip.ts:30]: This is a good candidate to be a [ternary](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_Operator).

```suggestion
        reply = Number(skipOption) > context.soundQueue.length
          ? All sounds skipped (count option was >= number of current sounds.)'
          : `Skipped ${ skipOption } sound(s).`
```

File: src/bot/commands/skip.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+      type: 'STRING',
+      required: false,
+      description: 'Enter amount to skip (including currently playing) or "all" to clear the queue entirely.',
+    }];
+  }
+
+  execute(interaction: CommandInteraction, context: BotContext): Promise<any> {
+    let reply;
+    const skipOption = interaction.options.getString(skipCommandOption, false);
+    if (context.botAudioPlayer.player.state.status === 'idle' && context.soundQueue.length === 0) { // If player is Idle
```
Comment (UncleDave) [inline src/bot/commands/skip.ts:20]: I'd like to encapsulate more of the `AudioPlayer` related logic. If we're going to wrap it in our own `BotAudioPlayer` then I wouldn't expect to interact with the internet `AudioPlayer`, I'd rather if `BotAudioPlayer` had a `stop` method and an `isIdle` method (or maybe even a property with a getter 😱). Ideally I'd like `botAudioPlayer.player` to be `private`.

File: src/bot/commands/skip.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        context.soundQueue.clear();
+        reply = 'Skipped all sounds. Queue is now empty.';
+      } else if (Number.isInteger(Number(skipOption))) {
+        if (Number(skipOption) > context.soundQueue.length) reply = 'All sounds skipped (count option was >= number of current sounds.)';
+        else reply = `Skipped ${ skipOption } sound(s).`;
+        context.botAudioPlayer.player.stop();
+        const int = Number(skipOption);
+        context.soundQueue.splice(0, int - 1);
+      } else {
+        reply = 'Invalid value for count entered. Try an integer or "all" without quotes.';
```
Comment (UncleDave) [inline src/bot/commands/skip.ts:35]: Only programmers and mathematicians use `integer` on a daily basis.

```suggestion
        reply = 'Invalid value for count entered. Try a whole number or "all" without quotes.';
```

**Review by UncleDave — CHANGES_REQUESTED** [review]

File: src/bot/bot-context.ts [hunk 5]
```diff
@@ -23,7 +25,21 @@ class SoundQueue extends Array<SoundQueueItem> {
...
+    this.player.play(resource);
+    return new Promise(resolve => {
+      this.player.on(AudioPlayerStatus.Idle, resolve);
+    });
+  }
+}
 
 export default class BotContext {
   soundQueue = new SoundQueue();
+
```
[the remark this answers is not in the document]
Comment (stuckinab0x) [reply src/bot/bot-context.ts:43]: > ESLint
Yes, all class members have COVID.

[the remark this answers is not in the document]
Comment (UncleDave) [reply src/bot/bot-context.ts:43]: Social distancing is an entirely reasonable excuse.

File: src/bot/commands/skip.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+      description: 'Enter amount to skip (including currently playing) or "all" to clear the queue entirely.',
+    }];
+  }
+
+  execute(interaction: CommandInteraction, context: BotContext): Promise<any> {
+    const skipOption = interaction.options.getString(skipCommandOption, false);
+    const count = Number(skipOption);
+    if (context.botAudioPlayer.state === 'idle' && context.soundQueue.length === 0)
+      // eslint-disable-next-line no-useless-escape
+      return interaction.reply({ content: 'No sounds currently playing or in queue! Why not try "/sound limmy are you deaf" ? \:ear_with_hearing_aid: \:smile:', ephemeral: true });
```
Comment (UncleDave) [inline src/bot/commands/skip.ts:22]: What is ESLint whinging about here? The `\:` bits? Does discord need those?

File: src/bot/commands/skip.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+      name: skipCommandOption,
+      type: 'STRING',
+      required: false,
+      description: 'Enter amount to skip (including currently playing) or "all" to clear the queue entirely.',
+    }];
+  }
+
+  execute(interaction: CommandInteraction, context: BotContext): Promise<any> {
+    const skipOption = interaction.options.getString(skipCommandOption, false);
+    const count = Number(skipOption);
```
Comment (UncleDave) [inline src/bot/commands/skip.ts:19]: Minor optimisation: move this down to where it's actually needed (line 32).

File: src/bot/commands/skip.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    this.commandData.options = [{
+      name: skipCommandOption,
+      type: 'STRING',
+      required: false,
+      description: 'Enter amount to skip (including currently playing) or "all" to clear the queue entirely.',
+    }];
+  }
+
+  execute(interaction: CommandInteraction, context: BotContext): Promise<any> {
+    const skipOption = interaction.options.getString(skipCommandOption, false);
```
Comment (UncleDave) [inline src/bot/commands/skip.ts:18]:
```suggestion
    const skipOption = interaction.options.getString(skipCommandOption);
```

The second parameter is optional and defaults to false so we might aswell omit it.

**Review by UncleDave — CHANGES_REQUESTED** [review]

File: src/bot/commands/skip.ts (hunk 6 above)
Comment (stuckinab0x) [reply src/bot/commands/skip.ts:22]: It's the `\` but it doesn't need them, I misread the discord.js guide. It's weird cause what it should actually have done is escaped the emoji but it's doing the same thing both ways.

We can ditch the `\`s and the ESLint comment then.

**Review by UncleDave — APPROVED** [review]

**Pull request merged**